### PR TITLE
Add role to deploy cASO

### PR DESCRIPTION
**Pull request by jovial**
cASO is an OpenStack accounting extractor. For more detail see:
https://github.com/IFCA/caso

By default, cASO is configured to output to Fluentd via TCP. The
accounting information can then be shipped off to ElasticSearch.

(cherry picked from commit d8ab00fa3c6ec3aba440ac50beb32c8e18e42337)
(cherry picked from commit 67ee60e807b1a85b92867eea69c888f132965cf0)
Change-Id: Ib81ec443bbb14d494d7c81dff617b371a1b68c52

### CODE REVIEW

File: ansible/roles/common/templates/conf/output/01-es.conf.j2 [hunk 1]
```diff
@@ -1,3 +1,19 @@
+{% if enable_caso | bool and inventory_hostname in groups['caso'] %}
+<match apel.events>
+    @type copy
+    <store>
+       @type elasticsearch
+       host {% raw %}{{ elasticsearch_address }}
+{% endraw %}
+       port {% raw %}{{ elasticsearch_port }}
+{% endraw %}
```
Comment (markgoddard) [inline ansible/roles/common/templates/conf/output/01-es.conf.j2:9]: Are you sure these should be raw?

Comment (markgoddard) [inline ansible/roles/common/templates/conf/output/01-es.conf.j2:9]: Are you sure these should be raw?

**Comment by priteau** [comment]
I guess it will fail if the inventory is missing the `caso` group?

**Comment by priteau** [comment]
I guess it will fail if the inventory is missing the `caso` group?

**Comment by priteau** [comment]
If this was an upstream patch I would point out the lack of `extra_volumes` support, but this could be merged without it.

**Comment by jovial** [comment]
> I guess it will fail if the inventory is missing the `caso` group?

Good point. I think I just need to add stuff like:

```
groups['caso'] | default([])
```
to any of the tasks outside of the caso role.

**Comment by priteau** [comment]
> > I guess it will fail if the inventory is missing the `caso` group?
> 
> Good point. I think I just need to add stuff like:
> 
> ```
> groups['caso'] | default([])
> ```
> 
> to any of the tasks outside of the caso role.

You can check how it was done for the libvirt-exporter change on Xena:

* https://review.opendev.org/c/openstack/kolla-ansible/+/831850
* https://review.opendev.org/c/openstack/kolla-ansible/+/833167
* https://review.opendev.org/c/openstack/kolla-ansible/+/834620

File: ansible/roles/common/templates/conf/output/01-es.conf.j2 (hunk 1 above)
Comment (jovial) [reply ansible/roles/common/templates/conf/output/01-es.conf.j2:9]: Hmm, good point. This does look wrong, I just copied this across from the original patch.

**Comment by priteau** [comment]
> > > > I guess it will fail if the inventory is missing the `caso` group?
> > > 
> > > 
> > > Good point. I think I just need to add stuff like:
> > > ```
> > > groups['caso'] | default([])
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > to any of the tasks outside of the caso role.
> > 
> > 
> > You can check how it was done for the libvirt-exporter change on Xena:
> > 
> > * https://review.opendev.org/c/openstack/kolla-ansible/+/831850
> > * https://review.opendev.org/c/openstack/kolla-ansible/+/833167
> > * https://review.opendev.org/c/openstack/kolla-ansible/+/834620
> 
> Wouldn't we never get there, since the play would not match in site.yml?

Good point, it's a new role rather than an extension of an existing one.

**Comment by jovial** [comment]
> > > > I guess it will fail if the inventory is missing the `caso` group?
> > > 
> > > 
> > > Good point. I think I just need to add stuff like:
> > > ```
> > > groups['caso'] | default([])
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > to any of the tasks outside of the caso role.
> > 
> > 
> > You can check how it was done for the libvirt-exporter change on Xena:
> > 
> > * https://review.opendev.org/c/openstack/kolla-ansible/+/831850
> > * https://review.opendev.org/c/openstack/kolla-ansible/+/833167
> > * https://review.opendev.org/c/openstack/kolla-ansible/+/834620
> 
> Wouldn't we never get there, since the play would not match in site.yml?

I think this would fail though: https://github.com/stackhpc/kolla-ansible/pull/205/files#diff-d33e12c3e3423adcb3eca3138dbc35ff4a5c224a307f38d088fce52a1ca9b56aR77

as it is outside the role?

File: ansible/roles/common/templates/conf/output/01-es.conf.j2 (hunk 1 above)
Comment (jovial) [reply ansible/roles/common/templates/conf/output/01-es.conf.j2:9]: The bizaare thing is that this works as is:
```
<match apel.events>
    @type copy
    <store>
       @type elasticsearch
       host 192.168.33.2
       port 9200
       logstash_format true
       logstash_prefix apel
       flush_interval 15s
       reload_connections false
       reconnect_on_error true
    </store>
</match>
```

Comment (jovial) [reply ansible/roles/common/templates/conf/output/01-es.conf.j2:9]: This is without:
```
# Included from conf/input/99-caso.conf.j2:
<source>
  @type tcp
  tag apel.events
  port 24224
  bind 127.0.0.1
  format /^(?<Payload>.*)$/
  emit_unmatched_lines true
</source>
```

**Comment by jovial** [comment]
I've removed the unnecessary raw tags and confirmed that a missing caso group doesn't break any tasks (provided that you do not set `enable_caso: true`.

**Review by markgoddard — APPROVED** [review]

**Pull request merged**